### PR TITLE
Switching from Alpine Image to multistaged opensuse/leap 

### LINKED ISSUE #2299: [Feature request] Improve Docker image performance with an alternative base image instead of Alpine
**Which version are you referring to**
3.1dev

**Please check this repo whether this is a known feature request**
No existing issue came up after a few searches and browsing of results. The closest I found were [some 2019 observations with an Alpine image](https://github.com/drwetter/testssl.sh/issues/1225).

**Describe your feature request (if it's a technical feature)**
It'd be nice if a Debian base image could be offered. It does not have the minimal disk usage that an alpine base image offers, but performance can be notably better.

**If your feature request otherwise is related to a usage problem, please describe it**
There are reports about DNS performance being problematic on Alpine images, but I did not notice a difference compared to testing against an IP address directly. So I assume it's more to do with `musl` vs `glibc`.

I have observed a repeatable difference between the current Alpine image and a Debian base image alternative of about 50% in time to perform a test with `--preference`, and alpine being 2-3x slower without `--preference`. The Debian image performed at native (outside of docker) or very close to.

**Describe the solution you'd like**

I adapted the current alpine Dockerfile to one for Debian:

```Dockerfile
# syntax=docker.io/docker/dockerfile:1
# HereDoc (EOF) feature (avoids needing `&& \`) requires BuildKit:
# https://docs.docker.com/engine/reference/builder/#here-documents

FROM debian:bullseye-slim

ARG DEBIAN_FRONTEND=noninteractive
RUN <<EOF
  ## Prepare package install
  apt-get -qq update  
  apt-get -qq upgrade  

  # Install packages
  apt-get install -qq --no-install-recommends bsdmainutils procps dnsutils

  # Cleanup
  apt-get -qq clean  
  rm -rf /var/lib/apt/lists/*  

  # Create testssl user
  adduser --system --group --uid 1000 --gecos "testssl user" --shell /bin/bash --disabled-password testssl

  # Prepare for COPY
  ln -s /home/testssl/testssl.sh /usr/local/bin/  
  mkdir -m 755 -p /home/testssl/etc /home/testssl/bin
EOF

USER testssl
WORKDIR /home/testssl/

COPY --chown=testssl:testssl etc/ /home/testssl/etc/
COPY --chown=testssl:testssl bin/ /home/testssl/bin/
COPY --chown=testssl:testssl testssl.sh /home/testssl/

ENTRYPOINT ["testssl.sh"]

CMD ["--help"]
```

- Image weight 138MB vs 41MB of current Alpine image.
- I have also tried with Fedora as another base image where performance was likewise better. It's very likely a `glibc` vs `musl` issue or something else going on with Alpine base images.
- There is also [this caveat mentioned about DNS on alpine](https://github.com/gliderlabs/docker-alpine/blob/master/docs/caveats.md#dns). A [related DNS issue implies it sends out A + AAAA record queries in parallel](https://github.com/gliderlabs/docker-alpine/issues/313#issuecomment-409872142) which apparently may be a cause.

---

## Comparison

Here are some observations I noticed when testing against a mailserver container (_with custom network and also with host network, not performance change_), with a docker image and native install of `testssl.sh:3.1dev`.

### Single test

```bash
# ~60 seconds to complete, same as native. Alpine ~150s:
$ docker run --rm -it --network test-network testssl:debian example.test:465

# 11-15 seconds to complete, likewise competitive with native (10-13s). Alpine 25-37s:
$ docker run --rm -it --network test-network testssl:debian example.test:465
```

### Parallel tests

These results are from running against the docker-mailserver test suite:

- `--quiet --preference --mode parallel` options are used, along with the ENV `ADDTL_CA_FILES` to verify chain of trust.
- Time in milliseconds is reported at the end of each test case.
- Each test case has the mail-server offer RSA, ECDSA, or both as the certificate, and with one of two cipher suite configs.
- Each test case is running `testssl.sh` with `--file` to perform tests across 7 ports from two services (Postfix and Dovecot):
  
  ```
  --jsonfile-pretty port_25.json --starttls smtp example.test:25
  --jsonfile-pretty port_587.json --starttls smtp example.test:587
  --jsonfile-pretty port_465.json example.test:465
  
  --jsonfile-pretty port_110.json --starttls pop3 example.test:110
  --jsonfile-pretty port_995.json example.test:995
  
  --jsonfile-pretty port_143.json --starttls imap example.test:143
  --jsonfile-pretty port_993.json example.test:993
  ```

```console
# Debian base image:
$ make clean generate-accounts test/tls_cipherlists

tls_cipherlists.bats
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=intermediate' + RSA [45863]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=intermediate' + ECDSA [36584]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=intermediate' + ECDSA with RSA fallback [55790]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=modern' + RSA [43647]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=modern' + ECDSA [36443]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=modern' + ECDSA with RSA fallback [52742]

6 tests, 0 failures in 274 seconds

# Alpine base image (worse case was around 330 sec):
$ make clean generate-accounts test/tls_cipherlists

tls_cipherlists.bats
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=intermediate' + RSA [52541]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=intermediate' + ECDSA [42963]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=intermediate' + ECDSA with RSA fallback [62151]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=modern' + RSA [45966]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=modern' + ECDSA [45853]
 ✓ [Security - TLS] (cipher lists) 'TLS_LEVEL=modern' + ECDSA with RSA fallback [57875]

6 tests, 0 failures in 311 seconds
```

The benefit was less apparent with `--mode parallel`, about 40 seconds or so difference. That is on a system with an old i5-6500 (4 cores / 4 threads) that would be under full load by either base image while running.

### LINKED ISSUE #2314: [Feature request] Dockerfile - Include `openssl1.1-compat`
Extracted from https://github.com/drwetter/testssl.sh/pull/2305 (_specifically [this commit](https://github.com/drwetter/testssl.sh/pull/2305/commits/ec33e4c1f2a626b3572d3f7c00d9081f489f8e25)_)

**Which version are you referring to**
Both releases: `3.0.x` + `3.1dev`

**Please check this repo whether this is a known feature request**
This is for the recent upgrade to Alpine 3.17 where OpenSSL package is now the `3.x` series, instead of the previous `1.1` series.
- There [was an attempt to PR this](https://github.com/drwetter/testssl.sh/pull/2305). It was closed for proposing too many changes at once.
- [An observation with local performance](https://github.com/drwetter/testssl.sh/issues/2299#issuecomment-1409545691) of `1.1` performing notably better than `3.x` for a CI test suite.

**If your feature request otherwise is related to a usage problem, please describe it**
- It was observed to be about 30 seconds faster (_~190 sec vs ~220 sec_). About a 15% improvement in run time.
- There may be other reasons to carry both versions of OpenSSL packages in the image, at least until the `3.x` series is more broadly available (_Debian Bullseye for example deploys with `1.1`_).

**Describe the solution you'd like**
- Adding the `openssl1.1-compat` package (adds 4MB).
- Alternatively offering a Debian or similar base image variant for `glibc` where performance seems better than Alpine with `musl` at the expense of image size.

---

This is just a suggestion based on insights from the referenced issue and closed PR.

It might be worth waiting to see if it interests / impacts anyone else, or if they also get approx 15% faster run times (_both for parallel and single tests_).

### LINKED ISSUE #2315: [Feature request] Dockerfile - Removing `git` package
Extracted from https://github.com/drwetter/testssl.sh/pull/2305 (_specifically [this commit](https://github.com/drwetter/testssl.sh/pull/2305/commits/ec33e4c1f2a626b3572d3f7c00d9081f489f8e25)_)

**Which version are you referring to**
Both releases: `3.0.x` + `3.1dev`

**Please check this repo whether this is a known feature request**
There [was an attempt to PR this](https://github.com/drwetter/testssl.sh/pull/2305). It was closed for proposing too many changes at once.

**Describe your feature request (if it's a technical feature)**
If the package isn't serving a purpose (_except for the similar `Dockerfile.git`_), it would reduce the image weight (_which was the [main motivation for choosing Alpine](https://github.com/drwetter/testssl.sh/commit/cbe38cc4bba6004d3cfd07c58a39fe239453de9d)_).

**Describe the solution you'd like**
Removing a package if it's not beneficial to the image. Reducing the image size by approx 7MB (19%).

---

This is just a suggestion based on insights from the closed PR. I'd raise a PR for this, but there is already one active touching that line, so I am documenting the suggestion here.

### LINKED ISSUE #2329: BUG: error querying OCSP responder
**Before you open an issue please check which version you are running and whether it is the latest in stable / dev branch**
I am running version 3.0.8

**Before you open an issue please whether this is a known problem by searching the issues**
I couldn't find anything 

**Command line / docker command to reproduce**
docker run --rm -ti drwetter/testssl.sh:3.0 --phone-out -S testssl.sh

**Expected behavior**
The following error message is not displayed
``` OCSP URI                     http://r3.o.lencr.org, error querying OCSP responder```

**Your system (please complete the following information):**
 - OS (outside of Docker): ``Ubuntu 20.04.5 LTS``
 - Platform (outside of Docker): ``Linux 5.4.0-94-generic x86_64`` 
 - OpenSSL (inside Docker): ``/home/testssl/bin/openssl.Linux.x86_64``

**Additional context**
The error message does not show up when omitting the ```--phone-out``` option, but I guess this is only suppressing the faulty OCSP request and not actually solving the issue. See ```docker run --rm -ti drwetter/testssl.sh:3.0 -S testssl.sh```

HTTP connections from inside the docker container to the OCSP URL seem to be working:
```
docker run --rm -ti --entrypoint /usr/bin/curl drwetter/testssl.sh:3.0  -v http://r3.o.lencr.org/
*   Trying 2.16.186.10:80...
* Connected to r3.o.lencr.org (2.16.186.10) port 80 (#0)
> GET / HTTP/1.1
> Host: r3.o.lencr.org
> User-Agent: curl/7.87.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Server: nginx
< Content-Length: 0
< Cache-Control: max-age=19229
< Expires: Wed, 08 Mar 2023 13:10:10 GMT
< Date: Wed, 08 Mar 2023 07:49:41 GMT
< Connection: keep-alive
< 
* Connection #0 to host r3.o.lencr.org left intact
```

Also no difference when using the latest DEV version:
```
docker pull drwetter/testssl.sh
Using default tag: latest
latest: Pulling from drwetter/testssl.sh
Digest: sha256:90e073332d15789047a22ccb3a65acaae7707c50a1877e8b054db5fe6289ee2b
Status: Image is up to date for drwetter/testssl.sh:latest
docker.io/drwetter/testssl.sh:latest

docker run --rm -ti drwetter/testssl.sh --phone-out -S testssl.sh
[..]
 Certificate Revocation List  --
 OCSP URI                     http://r3.o.lencr.org, error querying OCSP responder
 OCSP stapling                offered, not revoked
 OCSP must staple extension   --
[..]
```

Issue is persistent when running the above commands on a local notebook as well as on a virtual server in a data center.

### PR DESCRIPTION
**Pull request by drwetter**
See #2344 .

Also:

* fixes #2329 (minor bug)
* fixes #2315 (minor polish)
* fixes #2314 (feature / performance)
* fixes #2299 (performance)

24h to test before it gets merged.

### CODE REVIEW

**Comment by drwetter** [comment]
@polarathene : Care to update the Dockerfile.git too? 

If so, please commit to the suse_docker branch, so that it'll show up here. ``git`` is needed in that file?

File: CREDITS.md [hunk 1]
```diff
@@ -84,6 +84,9 @@ Full contribution, see git log.
 * Hubert Kario
   - helped with avoiding accidental TCP fragmentation
 
+* Brandon Kinney
```
Comment (polarathene) [inline CREDITS.md]:
```suggestion
* Brennan Kinney
```

**Comment by polarathene** [comment]
> @polarathene : Care to update the Dockerfile.git too?

I need to prioritize other tasks sorry :confounded: 

I'm not likely to have time to spare until April.

---

```Dockerfile
ARG BUILD_VERSION
ARG URL=https://github.com/drwetter/testssl.sh.git
RUN git clone --depth 1 --branch "${BUILD_VERSION}" "${URL}" /home/testssl
```

That is the main difference (_the `test -n` condition fails earlier, but if the ARG is empty `git` will still throw an error_).

- I am not sure what you need to offer this for vs just cloning locally to build the image if you're going to build one anyway? IIRC, it was primarily added for CI usage.
- Not an option for you, but there is `ADD` instruction support for [remote git repo sources](https://docs.docker.com/engine/reference/builder/#adding-a-git-repository-add-git-ref-dir) that doesn't requiring installing `git` into the container.
- You could also add this into the `builder` stage (not ideal), and add two more stages where they diverge at the `COPY` instruction, this would use `docker build --target stage-name-here` to decide what to build.

Personally, if your CI is the only known consumer of it, I'd advise adjusting your CI to provide the git repo branch instead.

That said, easiest fix is to just copy this `Dockerfile`, and place the 3 instructions above into the `builder` stage (_change the `/home/testssl` to `"S{INSTALL_ROOT}/home/testssl`_) and drop the last `COPY` in the final stage. I think that'll work, just extra maintenance to keep them in sync.

---

> `git` is needed in that file?

Yeah, you'd need to add that package if you want to keep that image variant.

File: CREDITS.md (hunk 1 above)
Comment (drwetter) [reply CREDITS.md]: Sorry!

**Comment by drwetter** [comment]
> Personally, if your CI is the only known consumer of it, 

probably not.

> I need to prioritize other tasks sorry

no prob. Thanks for your work so far!

**Pull request merged**